CenterForOpenScience/osf.io
# Update PreprintProvider to add Identifiers [OSF-7772]

## Purpose

We will now automatically create new DOIs (and ARK IDs while we're at it) for new preprints that are created via the OSF. This PR is to rearrange and add new methods that will make that possible.

When a preprint is published, a new DOI and ARK will be minted and saved as an Identifier. When that preprint is updated, updated metadata about the preprint will be sent to EZID and datacite.

## Changes

- Add the `IdentifierMixin` to `PreprintService` model, along with methods to set and create identifiers
- Tests for getting preprint identifiers
- Break out methods previously found in `project/views/register.py` and 
- Modify `get_identifier` method in the `IdentifierMixin` to work for both nodes and preprints
- Migration to add Identifiers to existing preprints
- Add mocking to `PreprintService` factory so that real IDs aren't created from the EZID service when the publish method is called
- When an update is made via the API and a DOI and ARK has already been minted, send those changes to the EZID service. This includes: 
    - privacy changes 
    - adding/removing contributors
    - deleting a preprint's node
- Test to make sure update method is called with preprints
- New `datacite_metadata_for_preprint` method to format metadata for version 4 of datacite's format
- Re-write `PrerpintFactory` so that most work happens in the `_create` method so that `_build` is truly without saving

## Side effects
Script that will be run to add DOIs and ARK ids to every existing Preprint. Without the script, everything should still work fine, but all new preprints will be issued a DOI and ARK id

## QA Considerations
- All new preprints should have 2 new identifiers associated with them
- These identifiers will be visible via the API, but *not* yet visible on the preprint detail page until Ember frontend work is done
- The Identifier on the API should appear as a relationship link
- The relationship should show two Identifiers - one DOI and one ARK
- The DOI will also appear in URL form under the "links" section on the detail (and list) view for preprints
- The doi's value should be in the format `add_me/osf.io/guid` and the ARK's value should be in the format ` `
- There is a script that needs to be run to back-fill DOIs and ARK ids for existing prerpints.
- Nothing should break before that script is run


## Ticket
Main ticket
https://openscience.atlassian.net/browse/OSF-7772

Other tickets:
https://openscience.atlassian.net/browse/OSF-7807
https://openscience.atlassian.net/browse/OSF-7817
